pkg/status/combinedstatus-resolution.go

```diff
@@ -0,0 +1,358 @@
+	// - MAX: the maximum value of the workstatuses groupBy values.
+	// If a combinedField's eval is nil, its TYPE is COUNT.
+	combinedFieldsEval map[string]ref.Val
+
+	selectEval map[string]ref.Val
+}
+
+// setCollectionDestinations sets the collection destinations of the
+// combinedstatus resolution.
+// The given set is expected not to be mutated during and after this call.
+// The function returns:
+//   - removedDestinations: a boolean indicating if one or more destinations
+//     were removed.
+//   - a set of destinations that were added.
+func (c *combinedStatusResolution) setCollectionDestinations(destinations sets.Set[string]) (bool, sets.Set[string]) {
+	c.Lock()
+	defer c.Unlock()
+
+	removedDestinations := c.collectionDestinations.Difference(destinations)
+	newDestinations := destinations.Difference(c.collectionDestinations)
+	// if nothing changed, return
+	if len(removedDestinations) == 0 && len(newDestinations) == 0 {
+		return false, nil
+	}
+
+	c.collectionDestinations = destinations
+	// trim the statuscollector data that are not relevant anymore
+	for _, data := range c.statusCollectorNameToData {
+		if data.workStatusRefToData == nil {
+			continue
+		}
+
+		for clusterName := range removedDestinations {
+			delete(data.workStatusRefToData, clusterName)
+		}
+	}
+
+	return len(removedDestinations) > 0, newDestinations
+}
+
+// setStatusCollectors sets ALL the statuscollectors relevant to the
+// combinedstatus resolution.
+// The given map is expected not to be mutated during this call, its values
+// contain valid expressions, and are immutable.
+// If a value is nil, it is assumed that the associated statuscollector has not
+// been processed.
+// The function returns a tuple (removedSome, addedSome):
+//
+// - removedSome: true if one or more statuscollectors were removed.
+//
+// - addedSome: true if one or more statuscollectors were added.
+func (c *combinedStatusResolution) setStatusCollectors(statusCollectorNameToSpec map[string]*v1alpha1.StatusCollectorSpec) (bool, bool) {
+	c.Lock()
+	defer c.Unlock()
+
+	removedSome, addedSome := false, false
+
+	// remove statuscollector data that are not relevant anymore and update the
+	// statuscollector data that are. If one of the latter is updated, mark it as added
+	for statusCollectorName, statusCollectorData := range c.statusCollectorNameToData {
+		statusCollectorSpec, ok := statusCollectorNameToSpec[statusCollectorName]
+		if !ok {
+			delete(c.statusCollectorNameToData, statusCollectorName)
+			removedSome = true
+			continue
+		}
+
+		if !statusCollectorSpecsMatch(statusCollectorData.StatusCollectorSpec, statusCollectorSpec) {
+			c.statusCollectorNameToData[statusCollectorName].StatusCollectorSpec = statusCollectorSpec
+			addedSome = true
+		}
+	}
+
+	// add new statuscollector data
+	for statusCollectorName, statusCollectorSpec := range statusCollectorNameToSpec {
+		if statusCollectorSpec == nil {
+			continue // not cached
```

Comment:
I think the code would be easier to understand if a `name: nil` entry in `statusCollectorNameToSpec` were given the same semantics as `name` not being in the map. That would mean the data structure has a state-based consistency predicate. Taking `name:nil` to mean "no change" makes the proper data structure contents depend on more history than just the latest information.

Reply:
done.